corehq/apps/users/management/commands/nphcda_find_mismatches.py

```diff
@@ -0,0 +1,151 @@
+import csv
+import re
+from collections import namedtuple
+from typing import Iterable, TypedDict
+
+from django.core.management.base import BaseCommand
+
+from corehq.apps.locations.models import SQLLocation
+from corehq.apps.users.models import CommCareUser
+
+
+Location = namedtuple('Location', 'id name code')
+country = Location('abc123', 'Nambia', 'nambia')
+location_cache = {
+    'nambia|foo|bar|baz': 'def456',
+}
+
+
+UserRecord = namedtuple('UserRecord', 'state lga ward settlement username')
+user_cache = {}
+
+
+UserCorrection = TypedDict('UserCorrection', {
+    'username': str,
+    'user_id': str,
+    'location_id 1': str,
+})
+
+
+class LocationError(ValueError):
+    pass
+
+
+class UserError(ValueError):
+    pass
+
+
+class UserLocationError(Exception):
+    def __init__(self, correction: UserCorrection, *args, **kwargs):
+        self.correction = correction
+        super().__init__(*args, **kwargs)
+
+
+class Command(BaseCommand):
+    help = """
+    Find mismatched users.
+
+    Uses the data in the CSV file to find mobile workers in the domain
+    who have been matched to the wrong location.
+    """
+
+    def add_arguments(self, parser):
+        parser.add_argument('domain')
+        parser.add_argument('users_csv')
+
+    def handle(self, domain, users_csv, *args, **options):
+        self.stdout.write('username,user_id,location_id 1')
+        for user in iter_users(users_csv):
+            # Find location
+            try:
+                state = get_location(domain, user.state, country)
+                lga = get_location(domain, user.lga, state)
+                ward = get_location(domain, user.ward, lga)
+                settlement = get_location(domain, user.settlement, ward)
+            except LocationError as err:
+                self.stderr.write(str(err))
+                continue
+
+            # Output corrections
+            try:
+                confirm_user_location(domain, user, settlement)
+            except UserError as err:
+                self.stderr.write(str(err))
+                continue
+            except UserLocationError as err:
+                row = ','.join([
+                    err.correction['username'],
+                    err.correction['user_id'],
+                    err.correction['location_id 1'],
+                ])
+                self.stdout.write(row)
+
+
+def iter_users(csv_filename: str) -> Iterable[UserRecord]:
+    username_re = re.compile(r'^[A-Z]{2}/[A-Z]{3}[0-9]+$')
+    last_full_row = {}
+    with open(csv_filename, 'r') as csv_file:
+        for row in csv.DictReader(csv_file):
+            # Skip blank rows
+            if not any(v for v in row.values()):
+                continue
+            # Replace blank values with their previous value
```

Comment:
Sometimes a value is implied, not entered, if it is the same as the value on last populated row.

Reply:
oh wow!